controllers/twophase/state_machine.go

```diff
@@ -0,0 +1,288 @@
+		lastStart = status.Experiment.StartTime.Time
+	}
+
+	defer func() {
+		m.Chaos.SetNextStart(nextStart)
+		m.Chaos.SetNextRecover(nextRecover)
+	}()
+
+	counter := 0
+	// nextStart is always after nextRecover
+	for {
+		if nextRecover.After(now) {
+			startTime = lastStart
+
+			return apply(ctx, m, v1alpha1.ExperimentPhaseRunning, startTime)
+		}
+
+		if nextStart.After(now) {
+			return noop(ctx, m, v1alpha1.ExperimentPhaseWaiting, now)
+		}
+
+		lastStart = nextStart
+		start, recover, err := m.IterateNextTime(nextStart, *duration)
+		if err != nil {
+			m.Log.Error(err, "failed to get the next start time and recover time")
+
+			return false, err
+		}
+
+		nextStart = *start
+		nextRecover = *recover
+
+		counter++
+		if counter > iterMax {
+			err = errors.Errorf("the number of iterations exceeded while resuming from pause with nextRecover(%s) nextStart(%s)", nextRecover, nextStart)
+			return false, err
+		}
+	}
+}
+
+// This method changes the phase of an object and do some side effects
+// There are 6 different phases, so there could be 6 * 6 = 36 branches
+func (m *chaosStateMachine) run(ctx context.Context, targetPhase v1alpha1.ExperimentPhase, now time.Time) (bool, error) {
+	currentPhase := m.Chaos.GetStatus().Experiment.Phase
+	m.Log.Info("change phase", "current phase", currentPhase, "target phase", targetPhase)
+
+	return phaseTransitionMap[currentPhase][targetPhase](ctx, m, targetPhase, now)
+}
+
+func (m *chaosStateMachine) Into(ctx context.Context, targetPhase v1alpha1.ExperimentPhase, now time.Time) error {
+	updated, err := m.run(ctx, targetPhase, now)
+	if err != nil {
+		m.Log.Error(err, "error while excuting state machine")
+	}
+
+	if updated {
+		updateError := m.Update(ctx, m.Chaos)
+		if updateError != nil {
+			m.Log.Error(err, "fail to update")
+
+			err = updateError
+		}
+	}
+
+	return err
+}
+
+func (m *chaosStateMachine) IterateNextTime(startTime time.Time, duration time.Duration) (*time.Time, *time.Time, error) {
+	scheduler := m.Chaos.GetScheduler()
+	if scheduler == nil {
+		return nil, nil, errors.Errorf("misdefined scheduler")
+	}
+	m.Log.Info("iterate nextStart and nextRecover", "startTime", startTime, "duration", duration, "scheduler", scheduler)
+	nextStart, err := utils.NextTime(*scheduler, startTime)
+
+	if err != nil {
+		m.Log.Error(err, "failed to get the next start time")
+		return nil, nil, err
+	}
+	nextRecover := startTime.Add(duration)
+
+	counter := 0
+	// if the duration is too long, `nextRecover` could be after `nextStart`
+	// we can jump over a start to make sure `nextRecover` is before `nextStart`
+	for nextRecover.After(*nextStart) {
+		nextStart, err = utils.NextTime(*scheduler, *nextStart)
+		if err != nil {
+			m.Log.Error(err, "failed to get the next start time")
+			return nil, nil, err
+		}
+
+		counter++
+		if counter > iterMax {
+			err = errors.Errorf("the number of iterations exceeded with nextRecover(%s) nextStart(%s)", nextRecover, nextStart)
+			return nil, nil, err
+		}
+	}
```

Comment:
I have a question here. Since we are trying to find the first start after `nextRecover`, and `nextRecover` is not changed in the loop, I think we could directly do `nextStart = utils.NextTime(*scheduler, nextRecover)` without iteration

Reply:
For a `cron` scheduler, `nextStart = utils.NextTime(*scheduler, nextRecover)` is enough. For `@every` scheduler, the answer is not correct :cry: 

The two kinds of scheduler is a disaster to the implementation :sob: .

Reply:
TLDR: I think the [loop in `resume`](https://github.com/chaos-mesh/chaos-mesh/pull/1221/files#diff-d396848e41155ba5b591cf171a8ab550845161333a956422f14c840b611d16bcR148) is essential but not this one

I made some categorized discussion (Chinglish! 🙈) for `@every` scheduler, and it seems it could work in this situation. I'm not sure if there's some miscalculation in my discussion...

There are two situations we need to use `IterateNextStart`: `apply` and `resume`. And to make it clear, let's say `NEXTSTART` is the real nextStart we want, same with `NEXTRECOVER`.
1. For [`apply`](https://github.com/chaos-mesh/chaos-mesh/pull/1221/files#diff-d396848e41155ba5b591cf171a8ab550845161333a956422f14c840b611d16bcR53), 
1.1. normally `startTime = now`, `nextStart = NEXTSTART`, `nextRecover = now + duration = NEXTRECOVER`. So would not enter this loop and returns. 
2.2. only when `duration > cycle time` should we iterate `nextStart` to the next one after `nextRecover`, in this case, `nextStart = utils.NextTime(*scheduler, nextRecover)` could works well. We will ignore this case for following discussion to keep it simple.
2. For [`resume`](https://github.com/chaos-mesh/chaos-mesh/pull/1221/files#diff-d396848e41155ba5b591cf171a8ab550845161333a956422f14c840b611d16bcR117), let's say a cycle = (first start,  second start]
2.1. when `nextRecover/nextStart` is in the same cycle with the resume time: one of them is after `now` so it would not enter `IterateNextStart` and return
2.2. when `nextRecover/nextStart` is sereval loops before us, if we pause for a long time. It would enter this loop, find the first `nextStart` after `nextRecover` and returns, and keep iterating in loop in `resume`. Since when we resume, `nextRecover/nextStart` are in the same loop, only apply `nextStart, err = utils.NextTime(*scheduler, *nextStart)` one time could be enough. So also, `nextStart = utils.NextTime(*scheduler, nextRecover)` could works.
2.3.when  `nextRecover/nextStart` is in the last cycle of resume time, after we iterate several 2.2 in `resume`. Go into `IterateNextStart`, we get `nextStart = NEXTSTART`, `nextRecover = NEXTRECOVER`, go into 2.1. and finish. No need to enter this loop and return.

So there is two situations (1.2 and 2.2) we need this loop and both of them could be replaced in this case.

Reply:
For example:

```
now := "12:10"
scheduler := "@every 5m"
duration := "6m"
```

Then we expected `nextRecover = "12:16"` and `nextStart = "12:20"`, but if we use `nextStart = utils.NextTime(*scheduler, nextRecover)`, it will be `12:21`.

Reply:
You're right I misunderstood what `scheduler.Next()` do for `@every`. Thanks for the explanation!